sstring: specialize uninitialize_string() and use resize_and_overwrite if available

tchaikov opened the pull request:
- sstring: specialize uninitialized_string() using 'if constexpr'
- sstring: use resize_and_overwrite if available
- sstring: add `sstring::resize_and_overwrite()`

tchaikov commented on include/seastar/core/sstring.hh:
```diff
@@ -592,18 +592,26 @@ public:
-    // FIXME: use __resize_default_init if available
-    ret.resize(size);
-    return ret;
+    if constexpr (internal::is_sstring<string_type>::value) {
```
for a minimal reproducer, please see https://godbolt.org/z/he8TT6fG5

tchaikov commented:
before this change
```
---------------------------------------------------------
Benchmark               Time             CPU   Iterations
---------------------------------------------------------
make_sstring          125 ns          125 ns      5624377
```

after this change
```
---------------------------------------------------------
Benchmark               Time             CPU   Iterations
---------------------------------------------------------
make_sstring         56.6 ns         56.6 ns     12349530
```

with test of
```c++
#include <string>

#include <benchmark/benchmark.h>
#include <seastar/core/sstring.hh>

using namespace seastar;

std::string_view foo = "foo";
std::string bar = "bar";
sstring zed = "zed";
const char* baz = "baz";

static void make_sstring(benchmark::State& state) {
    benchmark::ClobberMemory();
    for(auto _ : state) {
        auto s = make_sstring(foo, bar, zed, baz, "bah");
        benchmark::DoNotOptimize(s.data());
    }
}

BENCHMARK(make_sstring);

BENCHMARK_MAIN();
```

avikivity commented on include/seastar/core/sstring.hh:
```diff
@@ -592,18 +592,26 @@ public:
 template <typename char_type, typename Size, Size max_size, bool NulTerminate>
 basic_sstring<char_type, Size, max_size, NulTerminate> uninitialized_string(size_t size) {
-    using sstring_type = basic_sstring<char_type, Size, max_size, NulTerminate>;
-    return sstring_type(sstring_type::initialized_later(), size);
```
How does this work? A function with no return?

tchaikov replied:
ahh, i forgot to remove this function.

avikivity commented on include/seastar/core/sstring.hh:
```diff
@@ -604,8 +604,11 @@ string_type uninitialized_string(size_t size) {
-        // FIXME: use __resize_default_init if available
+#ifdef __cpp_lib_string_resize_and_overwrite
+        ret.resize_and_overwrite(size, [](string_type::value_type*, string_type::size_type) {});
+#else
```
This is nice. We should implement basic_sstring::resize_and_overwrite() too, sstring should follow std::string wherever possible.

tchaikov replied:
added. but i think `sstring::resize_and_overwrite(pre_calculated_size, op)` is less performant than a combination of 
```c++
auto s = uninitialized_string(pre_calculated_size);
op(s.data(), s.size());
```

avikivity replied:
I think the compiler will expand it to the same code.

tchaikov replied:
i doubt it. i think it depends on if the `sstring::resize_and_overwrite(pre_calculated_size, op)` is called on an instance of `sstring` with the right size. if we have something like
```c++
sstring s;
s.resize_and_overwrite(pre_calculated_size, op);
```
where `pre_calculated_size` is 16 which is greater than the default `max_size`. then we would have to reallocate the string.

avikivity replied:
There's nothing to reallocate, since a default-constructed string is empty. It just allocates it once, just like initialized_later.

tchaikov replied:
ahh, true. the internal buffer is not used at all. so no need to reallocate the bits in it. sorry for missing this and thank you for explaining it to me.

tchaikov commented:
@avikivity could you help re-review this change?

- dropped the specialization
- added `sstring::resize_and_overwrite()`

avikivity commented on include/seastar/core/sstring.hh:
```diff
@@ -320,7 +320,18 @@ public:
         return *this;
     }
-
+#if __cplusplus > 202002L
```
I see no need to restrict it.

avikivity commented on include/seastar/core/sstring.hh:
```diff
@@ -320,7 +320,18 @@ public:
+     *  @param op the function object used for setting the new content of the string
+     */
+    template<class Operation>
+    void resize_and_overwrite(size_t n, Operation op) {
```
Please add a constraint (but use SEASTAR_CONCEPT so it still works on C++17).

avikivity commented on include/seastar/core/sstring.hh:
```diff
@@ -320,7 +320,18 @@ public:
+    template<class Operation>
+    void resize_and_overwrite(size_t n, Operation op) {
+        resize(n);
+        resize(std::move(op)(data(), n));
```
This should be optimized to use initialized_later to avoid double-initializing.

Perhaps (pseudocode)

```
   if (capacity() < n) {
       operator=(sstring(initialized_later, n));
   }
   _size = n;
   resize(op(begin(), n));
}
```

tchaikov commented:
@avikivity all comments addressed

- dropped the guard of `#if __cplusplus > 202002L`
- avoided the unnecessary `resize()` call if the new size is smaller than the original size
- added type constraints
- added `assert()` to check if the returned size is greater than specified `count`.
